File: cli/src/commands/rename.ts
/**
 * mrsf rename — Update sidecar when a document is renamed.
 */

import type { Command } from "commander";
import chalk from "chalk";
import path from "node:path";
import fs from "node:fs/promises";
import { discoverSidecar, findWorkspaceRoot } from "../lib/discovery.js";
import { parseSidecar } from "../lib/parser.js";
import { writeSidecar } from "../lib/writer.js";

export function registerRename(program: Command): void {
  program
    .command("rename <old-document> <new-document>")
    .description(
      "Update a sidecar after its document has been renamed or moved",
    )
    .action(async (oldDoc: string, newDoc: string) => {
      const parentOpts = program.opts();
      const cwd = parentOpts.cwd ?? process.cwd();
      const quiet = parentOpts.quiet ?? false;
      const verbose = parentOpts.verbose ?? false;

      const oldDocPath = path.resolve(cwd, oldDoc);
      const newDocPath = path.resolve(cwd, newDoc);
      const root = await findWorkspaceRoot(cwd);

      // Find old sidecar
      const oldSidecarPath = await discoverSidecar(oldDocPath, { cwd: root ?? cwd });
      let doc;
      try {
        doc = await parseSidecar(oldSidecarPath);
      } catch {
        console.error(
          chalk.red(`Sidecar not found for ${oldDoc} at ${oldSidecarPath}`),
        );
        process.exit(1);
      }

      // Update document reference
      doc.document = path.basename(newDocPath);

      // Compute new sidecar path
      const newSidecarPath = await discoverSidecar(newDocPath, { cwd: root ?? cwd });

      // Ensure target directory exists
      await fs.mkdir(path.dirname(newSidecarPath), { recursive: true });

      // Write new sidecar
      await writeSidecar(newSidecarPath, doc);

      // Remove old sidecar if different location
      if (oldSidecarPath !== newSidecarPath) {
        try {
          await fs.unlink(oldSidecarPath);
        } catch {
          // Already gone or inaccessible
        }
      }

      if (!quiet) {
        console.log(
          chalk.green(`Renamed: ${oldSidecarPath} → ${newSidecarPath}`),
        );
      }
      if (verbose) {
        console.log(chalk.dim(`  document: ${doc.document}`));
        console.log(chalk.dim(`  comments: ${doc.comments.length}`));
      }
    });
}

File: python/src/mrsf/cli/rename.py
"""mrsf rename <old-document> <new-document> — rename a document and its sidecar."""

from __future__ import annotations

import os
import sys

import click

from ..discovery import discover_sidecar
from ..parser import parse_sidecar
from ..writer import write_sidecar
from .main import MrsfContext, pass_ctx


@click.command("rename")
@click.argument("old_document", type=click.Path())
@click.argument("new_document", type=click.Path())
@pass_ctx
def rename_cmd(ctx: MrsfContext, old_document: str, new_document: str) -> None:
    """Rename a document and update its sidecar."""
    old_path = (
        os.path.join(ctx.cwd, old_document) if not os.path.isabs(old_document) else old_document
    )
    new_path = (
        os.path.join(ctx.cwd, new_document) if not os.path.isabs(new_document) else new_document
    )

    # Find old sidecar
    try:
        old_sidecar = discover_sidecar(old_path, cwd=ctx.cwd, config_path=ctx.config)
    except Exception:
        click.echo(click.style(f"No sidecar found for {old_path}", fg="red"), err=True)
        sys.exit(1)

    if not os.path.exists(old_sidecar):
        click.echo(click.style(f"Sidecar not found: {old_sidecar}", fg="red"), err=True)
        sys.exit(1)

    doc = parse_sidecar(old_sidecar)
    doc.document = os.path.basename(new_path)

    # Compute new sidecar path
    new_sidecar = new_path + ".review.yaml"
    if old_sidecar.endswith(".review.json"):
        new_sidecar = new_path + ".review.json"

    # Create target directory if needed
    os.makedirs(os.path.dirname(new_sidecar), exist_ok=True)

    write_sidecar(new_sidecar, doc)

    # Remove old sidecar if path changed
    if os.path.abspath(old_sidecar) != os.path.abspath(new_sidecar):
        try:
            os.unlink(old_sidecar)
        except OSError:
            pass

    if not ctx.quiet:
        click.echo(f"Renamed: {old_sidecar} → {new_sidecar}")
